matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/token/HomeserverAccessTokenProvider.kt

```diff
@@ -16,13 +16,99 @@
         @SessionId private val sessionId: String,
-        private val sessionParamsStore: SessionParamsStore
+        private val sessionParamsStore: SessionParamsStore,
+        private val authenticationService: AuthenticationService,
+        private val globalErrorReceiver: GlobalErrorReceiver
 ) : AccessTokenProvider {
-    override fun getToken() = sessionParamsStore.get(sessionId)?.credentials?.accessToken
+
+    companion object {
+        private val mutex = Mutex()
+    }
+
+    object Constants {
+        /**
+        The time interval before the access token expires that we will start trying to refresh the token.
+        This avoids us having to block other users requests while the token refreshes.
+        Choosing a value larger than DEFAULT_DELAY_MILLIS + DEFAULT_LONG_POOL_TIMEOUT_SECONDS guarantees we will at least have attempted it before expiry.
+         */
+        const val PREEMPT_REFRESH_EXPIRATION_INTERVAL = 60000
+    }
+
+    override fun getToken(): String? {
+        var accessToken: String?
+        // We synchronise in a blocking fashion here so that when refresh is required, a single request becomes the leader.
+        // On successful refresh the leader saves the credential and the new access token then becomes available to other requests when they are unblocked.
+        // We should never send multiple refresh requests as refresh tokens are single-use(they rotate with a new one returned in the response).
+        // Mishandled via race conditions and we could become unauthenticated.
+        runBlocking {
+            mutex.withLock {
+                accessToken = verifyExpiryAndRefreshIfStale()
+            }
+        }
+        return accessToken
+    }
+
+    private suspend fun verifyExpiryAndRefreshIfStale(): String? {
+        val credentials = sessionParamsStore.get(sessionId)?.credentials ?: return null
+
+        if (credentials.refreshToken.isNullOrEmpty() ||
+                credentials.expiryTs == null ||
+                System.currentTimeMillis() < (credentials.expiryTs - Constants.PREEMPT_REFRESH_EXPIRATION_INTERVAL)) {
+            return credentials.accessToken
+        }
+
+        Timber.d("## Token Refresh: Refreshing access token...")
+
+        var result: RefreshResult? = null
+        try {
+            result = authenticationService
+                    .getRefreshWizard(sessionId)
+                    .refresh(credentials.refreshToken)
+        } catch (throwable: Throwable) {
+            val serverError = throwable
+                    .let { it as? HttpException }
+                    ?.let { it.toFailure(globalErrorReceiver) as? Failure.ServerError }
+            if (serverError != null) {
+                Timber.d("## Token Refresh: Failed to refresh access token. error: $serverError")
+                if (serverError.error.code == M_UNKNOWN_TOKEN || serverError.error.code == M_FORBIDDEN) {
+                    Timber.d("## Token Refresh: Requires logout.")
```

Comment:
throw the `twowable` again to propagate?

Reply:
👍 Yea haven't really worked out exactly how I'm gonna handle some of the error cases here. Will be working more on this.

Reply:
Error handling added.